### PR TITLE
Feature/step type class 1571

### LINKED ISSUE #1571: Make config for different step types into separate classes
Currently the config for steps are vary cluttered and difficult to read. This is because the different types requires different validation, as seen below: 

https://github.com/equinor/ert/blob/d1385bbcc5cc8abece2d7fb601c9d7ddda01c47b/ert3/config/_stages_config.py#L83-L112

Suggestion is to instead make different sub-classes for the different types, and let them handle the validation. 

POC: 

```python

# test.py
from typing import Union, List

from pydantic import BaseModel, Field


class BasicModel(BaseModel):
    name: str


class SubModelA(BasicModel):
    type: str = Field("A", const=True)
    some_property: int


class SubModelB(BasicModel):
    type: str = Field("B", const=True)
    some_other_property: str


class Model(BaseModel):
    l: List[Union[SubModelA, SubModelB]]


m = Model.parse_obj(
    {
        "l": [
            {"name": "a", "type": "A", "some_property": 3},
            {"name": "my_b", "type": "B", "some_other_property": "3"},
        ]
    }
)
assert isinstance(m.l[0], SubModelA)
print(m.l[0], type(m.l[0]))
assert isinstance(m.l[1], SubModelB)
print(m.l[1], type(m.l[1]))
```

```bash
$ python test.py
name='a' type='A' some_property=3 <class '__main__.SubModelA'>
name='my_b' type='B' some_other_property='3' <class '__main__.SubModelB'>
```

Interesting links:
https://github.com/samuelcolvin/pydantic/issues/503
https://github.com/samuelcolvin/pydantic/pull/469/files
https://stackoverflow.com/questions/58301364/pydantic-and-subclasses-of-abstract-class
https://github.com/samuelcolvin/pydantic/issues/619
https://github.com/samuelcolvin/pydantic/issues/265

### PR DESCRIPTION
**Pull request by sregales-TNO**
**Issue**
Resolves #1571 

**Approach**
Step class was completely removed and replaced with the Union of its subparts Unix and Function classes.

### CODE REVIEW

**Comment by sregales-TNO** [comment]
In config.__init__ I patch Step to be a Union type, but this should be refactored further to where function/methods calling config should not refer to Step but to Function or Step directly. 
example:
instead of this
`
def some_func(step_config: ert3.config.Step):
    if step_config.transportable_commands is not None:
        ...
`
it should change to
`
def some_func(step_config: ert3.config.Step):
    if isinstance(step_config, ert3.config.Unix):
        ...
`

File: ert3/config/__init__.py [hunk 1]
```diff
@@ -1,3 +1,7 @@
+from typing import Union
+
 from ert3.config._ensemble_config import load_ensemble_config, EnsembleConfig
-from ert3.config._stages_config import load_stages_config, StagesConfig, Step
+from ert3.config._stages_config import load_stages_config, StagesConfig, Function, Unix
 from ert3.config._experiment_config import load_experiment_config, ExperimentConfig
+
+Step = Union[Function, Unix]
```
Comment (jondequinor) [inline ert3/config/__init__.py:7]: missing newline
```suggestion
Step = Union[Function, Unix]

```

File: ert3/config/_stages_config.py [hunk 2]
```diff
@@ -80,42 +81,36 @@ class TransportableCommand(_StagesConfig):
...
-        if step.get("type") == "function":
-            if cmd_names:
-                raise ValueError("Commands defined for a function stage")
-            if script_lines:
-                raise ValueError("Scripts defined for a function stage")
-            if not step.get("function"):
-                raise ValueError("No function defined")
-        return step
+
+    # backwards compatible
```
Comment (jondequinor) [inline ert3/config/_stages_config.py:83]: What's the rationale for compatibility? Let's just break api now

Comment (sregales-TNO) [reply ert3/config/_stages_config.py]: This will need a documentation update on the structure of stages.yml

Comment (sregales-TNO) [reply ert3/config/_stages_config.py]: Is there documentation for the infrastructure of stages.yml?
At the moment our documentation just points to stages.yml examples, and that is already been updated.
Is this change to the examples good enough? 
If yes, then we can notify changes in release notes and be done

**Review by jondequinor — APPROVED** [review]
I think this is great. Thank you

**Pull request merged**